tests/nightly/model_backwards_compatibility_check/model_backwards_compat_inference.py

```diff
@@ -0,0 +1,149 @@
+# specific language governing permissions and limitations
+# under the License.
+
+from common import *
+
+def test_module_checkpoint_api():
+	model_name = 'module_checkpoint_api'
+	print ('Performing inference for model/API %s' %model_name)
+	data = download_data_from_s3(model_name)
+	if data is None:
+		print ('No data files found for %s' %model_name)
+		return
+
+	test_data = data['data']
+	test_label = data['labels']
+
+	data_iter = mx.io.NDArrayIter(test_data, test_label, batch_size=10)
+
+	## For each MXNet version that has the saved models
+	for folder in get_top_level_folders_in_bucket(s3, model_bucket_name):
+		print ('Fetching files for MXNet version : %s and model %s' %(folder, model_name))
+	        model_files = download_model_files_from_s3(model_name, folder)
+		if len(model_files) == 0:
+			print('No training files found for %s for MXNet version : %s'%(model_name, folder))
+			continue
+	        ## Load the model and perform inference
+	        loaded_model = get_module_api_model_definition()
+
+	        sym, arg_params, aux_params = mx.model.load_checkpoint(model_name, 1)
+	        loaded_model.bind(data_shapes=data_iter.provide_data, label_shapes=data_iter.provide_label)
+	        loaded_model.set_params(arg_params, aux_params)
+
+	        old_inference_results = load_inference_results(model_name)
+	        inference_results = loaded_model.predict(data_iter)
+	        ## Check whether they are equal or not ?
+	        assert_almost_equal(inference_results.asnumpy(), old_inference_results.asnumpy())
+	        clean_model_files(model_files, model_name)
+	        print ('=================================')
+	
+	print ('Assertion passed for model : %s' %model_name)
+	        
+
+def test_lenet_gluon_load_params_api():
+	model_name = 'lenet_gluon_save_params_api'
+	print ('Performing inference for model/API %s' %model_name)
+	## Get data from S3
+	data = download_data_from_s3(model_name)
+	if data is None:
+		print ('No data files found for %s' %model_name)
+		return
+	
+	test_data = data['data']
+
+	for folder in get_top_level_folders_in_bucket(s3, model_bucket_name):
+		print ('Fetching files for MXNet version : %s and model %s' %(folder, model_name))
+	        model_files = download_model_files_from_s3(model_name, folder)
+		if len(model_files) == 0:
+			print('No training files found for %s for MXNet version : %s'%(model_name, folder))
+			continue
+	        ## Load the model and perform inference
+	        loaded_model = Net()
+	        loaded_model.load_params(model_name+'-params')
+	        output = loaded_model(test_data)
+	        old_inference_results = mx.nd.load(model_name + '-inference')['inference']
+	        assert_almost_equal(old_inference_results.asnumpy(), output.asnumpy())
+	        clean_model_files(model_files, model_name)
+	        print ('=================================')
+	print ('Assertion passed for model : %s' %model_name)
+
+def test_lenet_gluon_hybrid_imports_api():
```

Comment:
Do we really need to make a test case for each individual model? Wouldn't it be possible to just loop through the S3 bucket and automatically determine all models that have to be run? That way, we can add things as required without copy&pasting this boilerplate.

Reply:
We would still need to instantiate the models and load them using different loading APIs. Hence the boiler plate code cannot be done away with. I have tried to keep the boiler plate as small as possible to perform the inference.